### ISSUE TITLE
Avoiding SIGSEGV during stack walking

### ISSUE THREAD

**Issue by Jongy** [jvm-profiling-tools/async-profiler#513]
`Engine::getNativeTrace` involves some heuristics during its stack walking: https://github.com/jvm-profiling-tools/async-profiler/blob/7d25210d2cd49e988ad3969ceace66225f49f264/src/engine.cpp#L51-L74
(the check of `valid_pc`, the `0x40000` check, the `bottom` check)

These can result in crashes, such as:
```
C  [libasyncProfiler.so+0x7c47d]  Engine::getNativeTrace(void*, int, void const**, int)+0x8d
C  [libasyncProfiler.so+0xa4a2d]  Profiler::getNativeTrace(Engine*, void*, ASGCT_CallFrame*, int)+0x3d
C  [libasyncProfiler.so+0xa59d9]  Profiler::recordSample(void*, unsigned long long, int, Event*)+0x179
C  [libasyncProfiler.so+0x7c3ae]  ITimer::signalHandler(int, siginfo*, void*)+0x3e
```

the crashing opcode and the source line (per my manual match-assmebly-to-source-code :sweat_smile:  ) are:
```
   7c47d:       4c 8b 62 08             mov    0x8(%rdx),%r12
```
and
```
((const void**)fp)[FRAME_PC_SLOT]
```

This might be a problem in the heuristic, or some JVM quirk (the SIGSEGV code is `SEGV_ACCERR` which is weird... a page with no read permissions? A guard page? Safepoint?).

I suppose that it can bypassed with another heuristic check, but I'm thinking of 2 other possible approaches that can bypass this issue (and all possible next ones):
* Catch the `SIGSEGV` and ignore it if we're currently in `getNativeTrace` (and in other functions which make "dangerous" accesses). This is inspired by how `copy_{to,from}_user` work (or used to work) in Linux - the kernel, upon a fault, would inspect the faulting opcode and if it's in some designated list of "fixup opcodes" - it would accept the fault gracefully, and jump to some handling code. So `getNativeTrace` could wrap its dangerous memory accesses with designated wrappers that'd return `NULL` should a fault happen. This requires catching the `SIGSEGV` signal which the JVM already handles, and I don't know how easy would it be to "hook" it. So, another solution might be...
* Read `/proc/self/maps` and avoid accessing areas which do not have the read bit set. This is obviously racy with other threads that may map/unmap/mprotect pages concurrently - but for the specific areas that async-profiler looks at (such as the current thread's stack, that has received the signal) I suppose it's less likely. This is less intrusive than the previous idea, but more heavy performance-wise.

**Comment by apangin** [comment]
The problem is understood. It should not happen for a native code that maintains the "normal" frame linkage, but there is a possibility that some native library that uses `fp` as a general-purpose register leaves some garbage on a stack, which looks like a valid stack pointer but actually isn't. The chances are quite low though: the "random" address should be within 256K of the current stack pointer, property aligned, but fall beyond the actual stack bottom.

I also thought about protecting AsyncGetCallTrace and the native stack walker with a signal handler (using setjmp/longjmp), but there is a couple of issues with that.
1. The JVM sets its own SIGSEGV handler, and it does not expect the library code to override it. There is no guarantee that the user's signal handler will work properly, unless the JVM is started with the special signal interposition library - [libjsig](https://docs.oracle.com/javase/8/docs/technotes/guides/vm/signal-chaining.html). The JVM may reset the handler at any time or even refuse to work. However, current HotSpot implementation permits a custom SIGSEGV handler, it only prints a warning. OpenJ9 also sets its own signal handler, but I haven't checked yet how it deals with third partly handlers.
2. I'm afraid that a custom SIGSEGV handler may mask a real programming bug, which will never be fixed.

For the above reasons, I don't consider the custom handler as the primary fix. However, it could be an optional feature (off by default), e.g. switched by another safemode bit.

Testing the accessible stack range would be a cleaner solution. This can be easily done on macOS using `pthread_get_stackaddr_np`/`pthread_get_stacksize_np`. Unfortunately, there are no such functions in glibc. The closest alternative - `pthread_getattr_np` - is not signal safe and is rather expensive.

HotSpot maintains its own stack limits, but only for threads managed by the JVM. And again, there are alternative JVMs. But at least for HotSpot, there is a cheap way to get the stack range.

Another option to check if the address is readable - is to use some syscall that accepts a userspace address. This needs to be done only once per page. The highest known accessible address can be then cached in a per-thread map.

So I can't name the perfect solution right now, but I think the approch to get/cache the accessible stack range is worth deeper investigation.

**Comment by Jongy** [comment]
> The JVM sets its own SIGSEGV handler, and it does not expect the library code to override it. There is no guarantee that the user's signal handler will work properly, unless the JVM is started with the special signal interposition library - libjsig. The JVM may reset the handler at any time or even refuse to work. However, current HotSpot implementation permits a custom SIGSEGV handler, it only prints a warning. OpenJ9 also sets its own signal handler, but I haven't checked yet how it deals with third partly handlers.

I was thinking I could call `signal` / `sigaction`, get the previous handler and then install an override which'll forward unhandled calls to the first. I understand from your words that the JVM might not be graceful about it.
I suppose "unhandled" might be complicated to define, which also leads to the next point...

> I'm afraid that a custom SIGSEGV handler may mask a real programming bug, which will never be fixed.

Starting with only the `getNativeTrace` call, we'll only handle signals that originate in our thread during the execution of `getNativeTrace`, or even more specifically, during the precise section of performing the one stack word read. All others will be forwarded to the JVM (and if they are not something the JVM owns, the process will crash, as it does now). This way I don't think anything else might be masked out (besides that one read, which we deem okay).

> So I can't name the perfect solution right now, but I think the approch to get/cache the accessible stack range is worth deeper investigation.

The idea to cache the lowest page of stack that's been used is interesting. It has the assumption that stacks don't shrink (or that we can query the stack range signal-safe), which I'm not sure holds true, but is definitely cleaner. If I get to investigate it more deeply, I will report here.

**Comment by Jongy** [comment]
@apangin , I have pushed a PoC of the first option to https://github.com/Jongy/async-profiler/tree/fault-handling.

I tested it with this DSO that I `LD_PRELOAD`ed into some dummy Java app:
```
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/mman.h>
#include <sched.h>


#define STACK_SIZE (1 << 20)

int thread_func(void *arg) {
    __asm(
        "movq %rsp, %rbp\n"
        "addq $0x30000, %rbp\n"
        "loop:\n"
        "jmp loop\n"
    );
}

__attribute__((constructor)) void init(void) {
    unsigned char *child_stack = mmap((void*)0x300000, STACK_SIZE, PROT_WRITE | PROT_READ, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
    clone(thread_func, child_stack + STACK_SIZE, CLONE_VM | CLONE_FS | CLONE_FILES | CLONE_SIGHAND | CLONE_THREAD | CLONE_SYSVSEM, NULL);
    printf("Started bad FP thread\n");
}
```

this DSO creates a thread which explicitly messes-up RBP, such that it points to an unmapped area, but is still considered legit for stack-walking per async-profiler's heuristics. Running a standard build of async-profiler on a Java process preloaded with this DSO - crashes immediately in `Engine::getNativeTrace`.

Then, trying with my branch (which is a PoC for this mechanism I named "fixups" - name is taken from the kernel, where it's also named this way) - I wrapped the "dangerous" access in `Engine::getNativeTrace` and now it looks like this:
```
         prev_fp = fp;
-        pc = stripPointer(((const void**)fp)[FRAME_PC_SLOT]);
+        FIXUP_BLOCK(
+            pc = stripPointer(((const void**)fp)[FRAME_PC_SLOT]),
+            goto out);
         fp = ((uintptr_t*)fp)[0];
```
if the code in the first argument to `FIXUP_BLOCK` causes a fault, then the code in the second argument executes (in this case, bails out from `Engine::getNativeTrace` and returns the trace it has collected thus far).

I added a print in the `out` label, which proves that this label gets jumped to multiple times during async-profiler's execution.

Do you think such a mechanism (which will wrap additional read-only, potentially-crashing parts of async-profiler) could be added as an optional feature, enabled with a flag to the profiler?

**Comment by apangin** [comment]
I understand that this is only a PoC, not the final PR, but the approach itself looks to me a bit overengineered.

Why not using `siglongjmp`, which seems to be a more portable equivalent of what you implemented by hand? Dealing with platform-specific registers manually looks hacky and questionable, as getcontext/setcontext are only guaranteed to work in pair. Furthermore, this will not work completely on Alpine Linux, a popular distribution for container images, because musl libc [does not implement](https://wiki.musl-libc.org/open-issues.html#ucontext.h) getcontext/setcontext.

The proposed implementation also raises the question of efficiency. Making a syscall, or even two, on every guarded read seems excessive. Contention on `fixup_arr` (threads will compete for exclusive access to the same cells) may cause a performance issue, too.

Regarding the safe native stack walking, I can think of the following approaches that solve the problem in an arguably better way:

1. Probe the unknown stack pages with `mremap` and cache the highest accessible address in a map. This is a clean and portable solution, but to be efficient, it needs a careful implementation of a concurrent thread-to-address map.
2. A hacky but very efficient way: detect the offset of `stackaddr`/`stacksize` members of `pthread_t` once, and then get the accessible stack range directly from `pthread_self`.
3. Using SIGSEGV handler in a different way. Read dangerous pointers via a special `safeFetch` function. SIGSEGV handler will detect a crash inside `safeFetch`, and move the instruction pointer to the next instruction, placing the error value (e.g. 0) in the return register. The stack walking will automatically stop at pc == 0. This approach requires no per-call setup, no per-thread state, and does not suffer from concurrency issues, and therefore is quite efficient.

**Comment by Jongy** [comment]
> Why not using `siglongjmp`, which seems to be a more portable equivalent of what you implemented by hand? Dealing with platform-specific registers manually looks hacky and questionable, as getcontext/setcontext are only guaranteed to work in pair. Furthermore, this will not work completely on Alpine Linux, a popular distribution for container images, because musl libc [does not implement](https://wiki.musl-libc.org/open-issues.html#ucontext.h) getcontext/setcontext.

To my understanding, `siglongjmp` is like `longjmp`, with the addition of restoring the process signal mask. We don't want to use it in a signal handler, because we want to return from the signal handler and only then jump back into the code at the saved context; while `siglongjmp` jumps immediately, so we remain in the signal handler's context.

I didn't know that `getcontext` et al. are not implemented in musl. We can change the code to use `setjmp` instead - actually, I originally wrote it that way, but decided to use `getcontext` because it makes more sense, as `getcontext` returns a struct matching the registers struct passed to the signal handler (while using `setjmp` requires some translation between `jmp_buf` to `ucontext_t`).

> The proposed implementation also raises the question of efficiency. Making a syscall, or even two, on every guarded read seems excessive. Contention on `fixup_arr` (threads will compete for exclusive access to the same cells) may cause a performance issue, too.

This is a PoC so I didn't think about performance all that much. From the top of my head, I can think of improvements for those 2 specific issues you've mentioned:
* We can avoid the syscall (`gettid`) by using `pthread_self()` instead of the TID as the key in `fixup_arr`. Or, at least, we can cache the TID inside `FIXUP_BLOCK` so there's just one syscall and not two.
* We can make use of the profiler's thread start / thread end hooks, to maintain a map of threads -> fixup blocks - as each thread needs just one. This removes any chance for contention.

> 
> Regarding the safe native stack walking, I can think of the following approaches that solve the problem in an arguably better way:
> 
> 1. Probe the unknown stack pages with `mremap` and cache the highest accessible address in a map. This is a clean and portable solution, but to be efficient, it needs a careful implementation of a concurrent thread-to-address map.
> 2. A hacky but very efficient way: detect the offset of `stackaddr`/`stacksize` members of `pthread_t` once, and then get the accessible stack range directly from `pthread_self`.
> 3. Using SIGSEGV handler in a different way. Read dangerous pointers via a special `safeFetch` function. SIGSEGV handler will detect a crash inside `safeFetch`, and move the instruction pointer to the next instruction, placing the error value (e.g. 0) in the return register. The stack walking will automatically stop at pc == 0. This approach requires no per-call setup, no per-thread state, and does not suffer from concurrency issues, and therefore is quite efficient.

I agree with you that for this specific issue of `Engine::getNativeTrace`, there are better & less complicated solutions (I specifically liked the `mremap` idea, and in general, the caching of "lowest known-to-be-mapped stack address").
As I wrote in my previous comment - my idea here is to provide a generic mechanism that could be used in other sites (another site I currently have in mind is the call to `AsyncGetCallTrace`, which I believe should be "read-only" and abort-able at any point, but you might prove me wrong here :sweat_smile: ). Additionally, this mechanism will be usable with ease, in any other site we might discover in the future.

I totally understand if you prefer to solve this specific site of `Engine::getNativeTrace` with other, less complex solutions; But what would you think about this fixups approach for other cases?

**Comment by apangin** [comment]
> We don't want to use it in a signal handler, because we want to return from the signal handler and only then jump back into the code at the saved context

But why? What can go wrong otherwise?

> But what would you think about this fixups approach for other cases?

I don't have other notable cases in mind. I also thought about `AsyncGetCallTrace`, but strictly speaking, this is a "foreign" code that we can't be sure about. Ideally, crash protection should be in the AGCT implementation itself, but unfortunately, there isn't.

AGCT is not strictly read-only, I know at least two counterexamples: 1) In debug builds of the JVM, the [stack-allocated marker](https://github.com/openjdk/jdk/blob/299022dfacbcb49e3bc5beca8ff9b1fca1101493/src/hotspot/share/prims/forte.cpp#L461) resets the state in a destructor (which is not called after the signal handler); 2) Even in product builds, stack walking [may update](https://github.com/openjdk/jdk/blob/299022dfacbcb49e3bc5beca8ff9b1fca1101493/src/hotspot/share/code/nmethod.cpp#L2221) PcDesc cache. If something wrong is put in this cache, it may fire later during normal stack walking. I don't claim the severity of these issues - just showing that AGCT isn't as innocent as it may seem, and therefore wrapping it in a signal handler is walking on a thin ice.

**Comment by Jongy** [comment]
> But why? What can go wrong otherwise?

I am under the impression that the kernel manages some context during the signal handler execution. If the signal handler never returns, because it simply jumps back to the fixup code - that kernel context doesn't get cleaned up. I don't have any specific sample from the kernel sources but I can think of some areas that could go wrong (for example - let's say that I'm using `sigaltstack`; if my handler never returns, how will the kernel know that the alt stack can be reused for the next signal?)

About AGCT - I see. I also don't have any other "read only" cases, so I suppose this generic SIGSEGV fixups idea could be put off for now...
I saw your commits in `safe-stack`, they are a fine and less complex solution, so let's go with that.

**Comment by apangin** [comment]
OK, I did more research on the topic, and here are my findings:
1. `mremap` doesn't really help. I found just one suitable way to proble multiple stack pages at once: `pwritev` to a dummy file descriptor (`memfd` or `shm_open`). Its performance is fine, but not good enough to use without caching. The cache must be concurrent, signal safe, and since thread IDs can be reused, the cache must somehow handle thread start and termination. This is doable, but involves extra complication.
2. The approach of getting `stackaddr`/`stacksize` is not feasible on Linux, because of different pthread implementations, and because a thread can be started bypassing pthreads completely, like in your example above.
3. Loading dangerous pointers via a special `SafeAccess` helper seems to be the easiest and reliable approach. I ended up [implementing this one](https://github.com/jvm-profiling-tools/async-profiler/commit/cf17e5efc3aa7cfe7c808f7381fabeb3841a3d49). The platform-dependent part is minimal. Tested with different versions of HotSpot - it works fine. Also works on Zing, but the JVM issues a warning message. The fix is already in master, will get in the nearest release.
